plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestLazyConnectionFactory.java

```diff
@@ -30,8 +33,13 @@ public class TestLazyConnectionFactory
     public void testNoConnectionIsCreated()
             throws Exception
     {
-        ConnectionFactory failingConnectionFactory = session -> {
-            throw new AssertionError("Expected no connection creation");
+        StatisticsAwareConnectionFactory failingConnectionFactory = new StatisticsAwareConnectionFactory(new MockConnectorFactory())
```

Comment:
Here we clearly see that NOT using Dependency Injection in tests leads to creating a chain of many instances of various classes that are not part of what's being tested. Previously, when we used the interface, it was easy to override it in-line.

I believe introducing a 'test' module that would have bindings for Mocks of these classes like (Stats, Retrying, etc.) would make the code more readable. But I'm not doing that change up-front, as I've learned a few times already that most of the maintainers don't like the code to go in that direction.

Reply:
This is the outcome of https://github.com/trinodb/trino/pull/19191#discussion_r1354884666. Previously it was generic and one could do any chain. Now the chain is fixed. This is good and bad. Maybe it is more good as it cannot be built in an non production use case.